pennylane/transforms/decompose.py

```diff
@@ -0,0 +1,120 @@
+def null_postprocessing(results):
+    """A postprocessing function returned by a transform that only converts the batch of results
+    into a result for a single ``QuantumTape``.
+    """
+    return results[0]
+
+
+@transform
+def apply_decomposition(tape, gate_set=None, gate_rules=None, max_expansion=None):
+    """Decomposes quantum circuit into a desired gate set.
+
+    Args:
+        tape (QuantumScript or QNode or Callable): a quantum circuit.
+        gate_set (set, optional): A set of decomposition gates. Defaults to None. If ``None``, gate set defaults to all operators not currently present.
+        gate_rules (Callable[qml.operation.Operator, bool], optional): A rule set for the gate set to follow. Defaults to None.
+        max_expansion (int, optional): The maximum depth of the expansion. Defaults to None.
+
+    Returns:
+        qnode (QNode) or quantum function (Callable) or tuple[List[QuantumScript], function]:
+
+        The decomposed circuit. The output type is explained in :func:`qml.transform <pennylane.transform>`.
+
+    .. seealso:: :func:`~.pennylane.devices.preprocess.decompose` for a transform that is designed particularly for restricted gate sets on specific device architectures.
+
+    **Examples:**
+
+    >>> tape = qml.tape.QuantumScript([qml.IsingXX(1.2, wires=(0,1))], [qml.expval(qml.Z(0))])
+    >>> batch, fn = apply_decomposition(tape, gate_set = {"CNOT", "RX", "RZ"})
+    >>> batch[0].circuit
+    [CNOT(wires=[0, 1]),
+    RX(1.2, wires=[0]),
+    CNOT(wires=[0, 1]),
+    expval(Z(0))]
+
+    >>> @partial(apply_decomposition, gate_rules = lambda obj: len(obj.wires) <= 2)
+    >>> @qml.qnode(device)
+    >>> def circuit():
+    >>>     qml.Toffoli(wires = range(3))
+    >>>
+    >>> print(qml.draw(circuit)())
+    0: ───────────╭●───────────╭●────╭●──T──╭●─┤
+    1: ────╭●─────│─────╭●─────│───T─╰X──T†─╰X─┤
+    2: ──H─╰X──T†─╰X──T─╰X──T†─╰X──T──H────────┤
+
+    """
+
+    if gate_set is None:
+        gate_set = set(qml.ops.qubit.__all__) - set(op.name for op in tape.operations)
+
+    if gate_rules is None:
+        gate_rules = lambda op: False
+
+    if isinstance(gate_set, str):
+        gate_set = set([gate_set])
+
+    if isinstance(gate_set, (list, tuple)):
+        gate_set = set(gate_set)
+
+    def decomposer(op):
+        return op.decomposition()
+
+    def stopping_condition(op):
+        if not op.has_decomposition:
+            if op.name not in gate_set:
+                warnings.warn(
+                    f"{op.name} has no supported decomposition and is not in the gate set.",
+                    UserWarning,
+                )
+            return True
+        return (op.name in gate_set) or gate_rules(op)
+
+    if all(stopping_condition(op) for op in tape.operations):
+        return (tape,), null_postprocessing
+
+    try:
+        new_ops = [
+            final_op
+            for op in tape.operations
+            for final_op in qml.devices.preprocess._operator_decomposition_gen(
```

Comment:
You will need to implement your own version of `_operator_decomposition_gen` for this - currently, unsupported operations still raise a DeviceError from here